cogs/tags/tags.py

```diff
@@ -119,9 +120,29 @@ def tag_decoder(obj):
         return TagAlias(**obj)
     return obj
 
+def roles_or_mod_or_permissions(all_server_roles: dict = {}, **perms):
+    async def predicate(ctx):
+        server = ctx.guild
+        mod_roles = [r.name.lower() for r in await ctx.bot.get_mod_roles(server)]
+        admin_roles = [r.name.lower() for r in await ctx.bot.get_admin_roles(server)]
+        roles = mod_roles + admin_roles + list(all_server_roles[server])
+        return await role_or_permissions(ctx, lambda r: r.name.lower() in roles, **perms)
+    return commands.check(predicate)
 
 class Tags(commands.Cog):
     """The tag related commands."""
+    allowed_roles = defaultdict(lambda: set([]))
+
+    def setAllowedRoles(self, guild: "guild ID", roles: list):
+        self.allowed_roles[guild] = set(roles)
+    def addAllowedRole(self, guild: "guild ID", role):
+        self.allowed_roles[guild].add(role)
+    def removeAllowedRole(self, guild: "guild ID", role):
+        self.allowed_roles[guild].discard(role)
+    def getAllowedRoles(self):
+        # tiers = await self.configV3.guild(ctx.guild).tiers()
+        # return tiers.keys()
+        return list(self.allowed_roles)
```

Comment:
I'm guessing syncAllowedRoles would work similarly to setAllowedRoles? I wanted to have add/remove functions because it seemed for efficient than overwriting the whole thing. We could also have allowed_roles be an object with set/get/add/remove methods.

Reply:
Yeah I was thinking the latter with saving to configV3 and then updating internal var. Efficiency-wise, yes you're right, but I was thinking since we're really only dealing with us as a server, it's okay since we rarely touch it anyways. In situations where the bot is home to multiple guilds, then yeah updating one at a time is good (I also think we might need a lock for it?).